lib/tdf3/src/tdf.ts

```diff
@@ -941,60 +941,64 @@ export class TDF extends EventEmitter {
     let encryptedOffset = 0;
 
     // eslint-disable-next-line @typescript-eslint/no-this-alias
-    const that = this;
-    const outputStream = makeStream({
-      async pull(controller: ReadableStreamDefaultController) {
-        while (
-          segments.length &&
-          Number.isInteger(controller.desiredSize) &&
-          (!controller.desiredSize || controller.desiredSize > 0)
-        ) {
-          const segment = segments.shift();
-          if (!segment) {
-            // Popped past the end
-            break;
-          }
-          const encryptedSegmentSize = segment.encryptedSegmentSize || encryptedSegmentSizeDefault;
-          const encryptedChunk = await zipReader.getPayloadSegment(
-            centralDirectory,
-            '0.payload',
-            encryptedOffset,
-            encryptedSegmentSize
-          );
-          encryptedOffset += encryptedSegmentSize;
-
-          // use the segment alg type if provided, otherwise use the root sig alg
-          const segmentIntegrityAlgorithmType =
-            that.manifest?.encryptionInformation.integrityInformation.segmentHashAlg;
-          const segmentHashStr = await that.getSignature(
-            reconstructedKeyBinary,
-            Binary.fromBuffer(encryptedChunk),
-            segmentIntegrityAlgorithmType || integrityAlgorithmType
-          );
+    // const that = this;
 
-          if (segment.hash !== base64.encode(segmentHashStr)) {
-            throw new ManifestIntegrityError('Failed integrity check on segment hash');
-          }
+    const underlyingSource = {
+      pull : async (controller: ReadableStreamDefaultController) => {
 
-          let decryptedSegment;
+        if (segments.length === 0) {
+          controller.close();
+          return;
+        }
 
-          try {
-            decryptedSegment = await cipher.decrypt(encryptedChunk, reconstructedKeyBinary);
-          } catch (e) {
-            throw new TdfDecryptError(
-              'Error decrypting payload. This suggests the key used to decrypt the payload is not correct.',
-              e
-            );
-          }
+        const segment = segments.shift();
+        if (!segment) {
+          throw new Error('Shifted past end of segments array');
+        }
+        const encryptedSegmentSize = segment?.encryptedSegmentSize || encryptedSegmentSizeDefault;
+        const encryptedChunk = await zipReader.getPayloadSegment(
+          centralDirectory,
+          '0.payload',
+          encryptedOffset,
+          encryptedSegmentSize
+        );
+        encryptedOffset += encryptedSegmentSize;
+
+        // use the segment alg type if provided, otherwise use the root sig alg
+        const segmentIntegrityAlgorithmType =
+          this.manifest?.encryptionInformation.integrityInformation.segmentHashAlg;
```

Comment:
Instead of cluttering up the code with `?.` save the manifest above and assert that it is present like you do with the `segment` variable.